queues.go

```diff
@@ -0,0 +1,60 @@
+package main
+
+// A queue must be read and write
+type Queue interface {
+	send(DomainEntry) error
+	pull(int) ([]DomainEntry, error)
+	peek(string) int
+	persist() error
+}
+
+type cachedQueue struct {
+	store   map[string]int
+	backend Backend
+}
+
+func NewCachedQueue(persistanceBackend Backend) *cachedQueue {
+	q := new(cachedQueue)
+	q.store = make(map[string]int)
+	q.backend = persistanceBackend
+	return q
+}
+
+func (q cachedQueue) send(entry DomainEntry) error {
+	stats := q.store[entry.Domain]
```

Comment:
Would you have enough memory to store millions of different domains? What if an OOM killer comes along?

Reply:
Go will return me a zeroed struct if the key is missing. Which is nice since the zero means that it is new. It will then add to it soon after. Alas this is tricky and deserves a comment or at least something more explicit.

Reply:
You could also potentially check for the existence of that key within the map :)

Reply:
Yes. Explicit better than implicit. Though I do worry about an addition branch step but that's some serious preoptimization when there are bigger fish to fry.